job/models.py

```diff
@@ -3350,6 +3331,689 @@ class GeneratorOutputs(BaseModel, models.Model):
+        validators=[
+            MinValueValidator(0),
+            MaxValueValidator(1e4)
+        ],
+        blank=True,
+        help_text="Pounds of CO2 emitted per MMBTU of CHP fuel burned."
+    )
+    emissions_factor_lb_PM25_per_mmbtu = models.FloatField(
+        validators=[
+            MinValueValidator(0),
+            MaxValueValidator(1e4)
+        ],
+        blank=True,
+        help_text="Pounds of CO2 emitted per MMBTU of CHP fuel burned."
+    )
+    
+    def clean(self):
+
+        self.fuel_cost_per_mmbtu = scalar_to_vector(self.fuel_cost_per_mmbtu)
+
+        if self.emissions_factor_lb_CO2_per_mmbtu == None:
+            self.emissions_factor_lb_CO2_per_mmbtu = FUEL_DEFAULTS["emissions_factor_lb_CO2_per_mmbtu"].get(self.fuel_type, 0.0)
+        
+        if self.emissions_factor_lb_SO2_per_mmbtu == None:
+            self.emissions_factor_lb_SO2_per_mmbtu = FUEL_DEFAULTS["emissions_factor_lb_SO2_per_mmbtu"].get(self.fuel_type, 0.0)
+        
+        if self.emissions_factor_lb_NOx_per_mmbtu == None:
+            self.emissions_factor_lb_NOx_per_mmbtu = FUEL_DEFAULTS["emissions_factor_lb_NOx_per_mmbtu"].get(self.fuel_type, 0.0)
+        
+        if self.emissions_factor_lb_PM25_per_mmbtu == None:
+            self.emissions_factor_lb_PM25_per_mmbtu = FUEL_DEFAULTS["emissions_factor_lb_PM25_per_mmbtu"].get(self.fuel_type, 0.0)
+
+        error_messages = {}
+
+        # possible sets for defining CHP
+        update_installed_costs = False
+        if (type(self.dict.get("installed_cost_per_kw")) == float or
+                (type(self.dict.get("installed_cost_per_kw")) == list and len(self.dict.get("installed_cost_per_kw")) == 1)
+                ):
+            self.tech_sizes_for_cost_curve = []
+        elif self.dict.get("installed_cost_per_kw") not in [None, [], ""] and self.dict.get("tech_sizes_for_cost_curve") in [None, [], ""]:
+            error_messages["required inputs"] = \
+                "To model CHP cost curve, you must provide `chp.tech_sizes_for_cost_curve` vector of equal length to `chp.installed_cost_per_kw`"
+        elif self.dict.get("tech_sizes_for_cost_curve") in [None, [], ""]:
+            update_installed_costs = True
+        elif len(self.dict.get("tech_sizes_for_cost_curve")) != len(self.dict.get("installed_cost_per_kw")):
+            error_messages["required inputs"] = \
+                "To model CHP cost curve, you must provide `chp.tech_sizes_for_cost_curve` vector of equal length to `chp.installed_cost_per_kw`"
+        
+        if not at_least_one_set(self.dict, self.possible_sets):
+            error_messages["required inputs"] = \
+                "Must provide at least one set of valid inputs from {}.".format(self.possible_sets)
+        else:  #use defaults if CHP prime mover given 
+            if self.dict.get("prime_mover") not in [None, ""]:
+                boiler_type = "steam" if self.prime_mover == "combustion_turbine" else "hot_water"
```

Comment:
@zolanaj I'm realizing now that this assignment of CHP defaults will have to go in validators.py because "steam" or "hot_water" could be a user-input for ExistingBoiler which would change this selection for thermal_efficiency. We also may in the future do the logic of size_class based on the average heating load, which would require the cross_clean functionality. Thoughts or downsides of moving this there?

Reply:
I also just realized this same logic for assigning CHP defaults is replicated in REopt.jl. Thoughts on the need to replicate this type of conditional logic in the API? The obvious benefit is that the django models get assigned the default value, whereas if we just relied on REopt.jl then the api_response would not include those assigned values in the inputs. I'm planning to integrate the API's chp_defaults view/endpoint logic into REopt.jl so that the `size_class` gets chosen based on the average heating load, so that would also have to be replicated in the API validators.py. It starts to get a bit complicated to be replicating in two places.

Reply:
FYI I'm going to work on creating a Julia function that will contain all of the API's `/chp_defaults` view logic and assign default values. So that will be used on the Julia-side, and I'll expose that function in `http.jl` as a Julia endpoint for validators to call so that we don't have to duplicate the logic. I think it shouldn't be a problem to call an `http.jl` endpoint in validators, right?

Reply:
@Bill-Becker sorry this took a while to respond.  I'm in favor of simplifying the API and doing the work in Julia as much as possible the way we do for financials.  If I'm understanding your method of exposing the Julia-side assignment of default values in `http.jl`, Julia would return the default values that would then update the inputs before `optimize!` is called to generate what becomes the outputs later - is that right?  If so, I think you found an elegant way of avoiding duplicate input handling and I support it.